internal/upload/service.go

```diff
@@ -48,49 +47,37 @@ func NewService(logger logr.Logger, config Config, uploadMaxSize int, tokenVerif
 		tokenVerifier: tokenVerifier,
 	}
 
-	err = s.openBucket(config)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+
+	err = s.openBucket(ctx, config)
```

Comment:
I think it's a mistake to open the bucket at this point - it's possible it's never accessed, and it hides the need to call `bucket.Close()` when the resource is no longer needed.  I think it's better to open the bucket in the service methods that need to access the contents.

Reply:
In general I think we are sloppy about closing buckets in Enduro.

Reply:
I believe that it's meant to be used like *http.Client or *sql.DB, i.e. they're meant to be shared, can be used by multiple goroutines, etc. The Close method closes the bucket. But I will look it up again. I think we were doing this in sdps_preprocessing if I remember correctly.

Reply:
The example at https://gocloud.dev/howto/blob/#opening includes an immediate `defer bucket.Close()` which makes me think there is a cost to keeping buckets open.

Reply:
should we add a lint check for bucket closing or closing and other structs with a close method?

Reply:
I've looked this up more closely and this is what I've found:
* A quick search on GitHub ([link](https://github.com/search?q=*blob.Bucket+language%3AGo&type=code)) reveals that it's a common practice to have a shared bucket as a struct field. Some examples: [goreleaser](https://github.com/goreleaser/goreleaser/blob/c9cc1065b473aa5ae96c4a0074b360a9f37e6f2b/internal/pipe/blob/upload.go#L233), [open-saves](https://github.com/googleforgames/open-saves/blob/536314ac9be11e64a3e642e213e9c8fa354b6d86/internal/pkg/blob/gcs.go#L31), [go-iiif](https://github.com/go-iiif/go-iiif/blob/6aba66ab5764973c0b1ef563d3c6ec8ec54d904a/source/memory.go#L13), [vroom](https://github.com/getsentry/vroom/blob/6ac471577674e6628e3aa2575e0f9be7f6a7a5c4/cmd/vroom/main.go#L41)... 
* `Close()` seems to be a no-op for most driver implementations, e.g.: [memblob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/memblob/memblob.go#L102-L104), [s3blob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/s3blob/s3blob.go#L370-L372), [gcsblob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/gcsblob/gcsblob.go#L400-L402) or [fileblob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/fileblob/fileblob.go#L293-L295). [*blob.Bucket](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/blob.go#L1276-L1286) (which performs as the wrapper in the driver pattern) is only wrapping the call to the underlying driver and seems like it's only preventing concurrent access (with a shared lock), but it does not seem to be performince other clean-up operations.
* Underlying drivers, e.g. [s3blob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/s3blob/s3blob.go#L362-L368) or [gcsblob](https://github.com/google/go-cloud/blob/5ec203b7dc558ebb940d0847f8dae12699f1d873/blob/gcsblob/gcsblob.go#L344), use their bucket struct to store the client objects which seem to be expensive to create, have thread-safety capabilities and generally seem to be designed to be long-lived. Opening a s3 or gcs bucket seems to be particularly expensive because the underlying SDKs need to deal with concerns like authentication or metadata, e.g. opening a gcs bucket results in network operations, i.e. see [Creating a client](https://pkg.go.dev/cloud.google.com/go/storage#hdr-Creating_a_Client) which reads: clients should be reused instead of created as needed. 
* I wrote a quick [benchmark](https://gist.github.com/sevein/856fc782803eebeb222a8be76193388d). If I'm intepreting it correctly I think it's showing that `BenchmarkWithOpen` allocates approximately 752 more bytes and performs 15 more allocations than `BenchmarkWithoutOpen`, and took 1038 ns more to run (lol), but this exacerbated considerably when using s3blob.

My understanding is that the cost of keeping a bucket open is negligible compared to the cost of opening the bucket each time it's needed. The bootstrap process appears to be more resource-intensive, so it's better to do it just once. Some memory is used for caching purposes, but this will save computation time later on.

Reply:
> should we add a lint check for bucket closing and other structs with a close method?

There is bodyclose or sqlclosecheck in golangci-lint, but I don't think there is a general-purpose linter probably because it's an application-specific concern, not a primitive in the language, if that makes sense.